File: vulntest/tier1-single/deserialization/go/source/main.go
// Tier 1 — insecure deserialisation (Go / encoding/gob).
//
// encoding/gob.Decoder on attacker-controlled stream. gob is Go's
// native binary serialisation; like other binary formats, no
// length cap by default — large allocations driven by attacker
// content.

package main

import (
	"bytes"
	"encoding/gob"
	"fmt"
	"io/ioutil"
	"os"
)

type Message struct {
	ID      uint32
	Payload []byte
}

func main() {
	if len(os.Args) < 2 {
		fmt.Fprintln(os.Stderr, "usage")
		os.Exit(1)
	}
	data, err := ioutil.ReadFile(os.Args[1])
	if err != nil { panic(err) }

	dec := gob.NewDecoder(bytes.NewReader(data))
	var m Message
	// sink: decode attacker-controlled bytes
	if err := dec.Decode(&m); err != nil { panic(err) }
	fmt.Printf("id=%d payload_len=%d\n", m.ID, len(m.Payload))
}

File: vulntest/tier1-single/deserialization/rust/source/src/main.rs
// Tier 1 — insecure deserialisation (Rust / serde+bincode).
//
// bincode::deserialize() with attacker-controlled bytes can crash
// the program (panic) and, depending on the target type, can
// trigger logic flaws — e.g., negative size in a length-prefixed
// type. Pure-Rust deserialisation cannot RCE in the way Java/.NET
// gadget chains do (no reflection-based constructor invocation),
// but it can produce DoS / logic compromise.

use serde::Deserialize;
use std::env;
use std::fs;

#[derive(Deserialize, Debug)]
struct Message {
    id: u32,
    payload: Vec<u8>,
}

fn main() {
    let args: Vec<String> = env::args().collect();
    if args.len() < 2 { return; }
    let bytes = fs::read(&args[1]).expect("read");
    // sink: deserialize attacker-controlled bytes with no max-length cap
    let msg: Message = bincode::deserialize(&bytes).expect("deserialize");
    println!("id={} payload_len={}", msg.id, msg.payload.len());
}
